pkg/network/usm/sharedlibraries/ebpf.go

```diff
@@ -126,19 +181,182 @@ func (e *EbpfProgram) Init() error {
-	return e.perfHandler
+// InitWithLibsets initializes the eBPF program. It is guaranteed to perform the initialization only if needed
+// For example, if the program is already initialized to listen for a certain libset, it will not reinitialize
+// the program to listen for the same libset. However, if the libsets are changed, it will reinitialize the program.
+func (e *EbpfProgram) InitWithLibsets(libsets ...Libset) error {
+	// Ensure we have all valid libsets, we don't want cryptic errors later
+	for _, libset := range libsets {
+		if !IsLibsetValid(libset) {
+			return fmt.Errorf("libset %s is not valid, ensure it is in the LibsetToLibSuffixes map", libset)
+		}
+	}
+
+	// Lock for the initialization variables
+	e.initMutex.Lock()
+	defer e.initMutex.Unlock()
+
+	// If the program is initialized, check if the libsets are already enabled
+	if e.isInitialized {
+		// If the libsets are already enabled, return, we have nothing to do
+		if e.areLibsetsAlreadyEnabled(libsets...) {
+			return nil
+		}
+
+		// If not, we need to reinitialize the eBPF program, so we stop it
+		// We use stopImpl to avoid changing the refcount
+		e.stopImpl()
+	}
+
+	// Add the libsets to the enabled libsets
+	for _, libset := range libsets {
+		e.enabledLibsets[libset] = struct{}{}
+	}
+
+	e.setupManagerAndPerfHandlers()
+
+	if err := e.loadProgram(); err != nil {
+		return fmt.Errorf("cannot load program: %w", err)
+	}
+
+	if err := e.start(); err != nil {
+		return fmt.Errorf("cannot start manager: %w", err)
+	}
+
+	e.isInitialized = true
+	return nil
 }
 
-// Stop stops the eBPF program
+func (e *EbpfProgram) start() error {
+	err := e.Manager.Start()
+	if err != nil {
+		return fmt.Errorf("cannot init manager: %w", err)
+	}
+
+	for libset, perfHandler := range e.perfHandlers {
+		e.wg.Add(1)
+		go func(ls Libset, ph *ddebpf.PerfHandler) {
+			defer e.wg.Done()
+
+			dataChan := ph.DataChannel()
+			lostChan := ph.LostChannel()
+			for {
+				select {
+				case <-e.done[ls]:
+					return
+				case event, ok := <-dataChan:
+					if !ok {
+						return
+					}
+					e.handleEvent(&event, ls)
+				case <-lostChan:
+					// Nothing to do in this case
+					break
+				}
+			}
+		}(libset, perfHandler)
+	}
+
+	return nil
+}
+
+func (e *EbpfProgram) handleEvent(event *ebpf.DataEvent, libset Libset) {
+	e.callbacksMutex.RLock()
+	defer func() {
+		event.Done()
+		e.callbacksMutex.RUnlock()
+	}()
+
+	libpath := ToLibPath(event.Data)
+	for callback := range e.callbacks[libset] {
+		// Not using a callback runner for now, as we don't have a lot of callbacks
+		(*callback)(libpath)
+	}
+}
+
+// CheckLibsetsEnabled checks if the eBPF program is enabled for the given libsets, returns an error if not
+// with the libsets that are not enabled
+func (e *EbpfProgram) CheckLibsetsEnabled(libsets ...Libset) error {
+	e.initMutex.Lock()
+	defer e.initMutex.Unlock()
+
+	var err error
+	for _, libset := range libsets {
+		if _, ok := e.enabledLibsets[libset]; !ok {
```

Comment:
I think all of them are protected by the same mutex, `initMutex`. It's true that it can be a bit confusing, I added a comment to `areLibsetsAlreadyEnabled` which is the only function that didn't have the mutex explicitly taken, I'm open to suggestions to make it better though.